syscall: add support for getting uid/gid/pid and related numbers

aykevl opened the pull request:
This is for compatibility with the existing syscall package and in fact
these functions are implemented in the baremetal syscall version. The
numbers will most likely be fake for WebAssembly but that's the job of
wasi-libc to sort out.

I'm not sure how to add tests for these, as they are normally unknown by
definition (unlike command line arguments and environment variables
which can be set by a test program).

See: https://github.com/tinygo-org/tinygo/issues/1886

fgsch commented:
fwiw, this is missing the functions in the os package to fix #1886.

aykevl commented:
Rebased to fix merge conflicts.

> fwiw, this is missing the functions in the os package to fix #1886.

What do you mean? It does define `Getuid`.

fgsch commented:
> What do you mean? It does define Getuid.

This is defining the functions under syscall, but the error in #1886 is in the os package.
In other words, they need to be defined under os as well, same as in the stdlib.

aykevl commented:
Eh, oops. Thank you for pointing that out. I'll submit a fix.

fgsch commented:
This can go in as is and the changes to os can happen in a separate PR IMO.

aykevl committed:
compiler: implement syscall.rawSyscallNoError in inline assembly

This makes it possible to call syscall.Getpid() on Linux, for example.
These syscalls never return an error so don't need any error checking.

aykevl committed:
os: implement process related functions

This commit implements various process related functions like
os.Getuid() and os.Getpid(). It also implements or improves this support
in the syscall package if it isn't available yet.

aykevl commented:
I have revamped this PR to also add support for these functions in the os package and added (limited) testing that they can in fact be compiled. This surfaced an issue on Linux which I've also fixed. So now, `os.Getuid` and related functions should at least compile on most platforms.

Unfortunately, it's hard to test these functions as they are sometimes either unavailable entirely (wasm, baremetal) or otherwise their values depend on the environment.

deadprogram commented:
Seems more solid now that at least some smoke testing is happening. Let's merge and keep going from here.
